Add FCU TRK/FPA Mode

theomessin opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #1470, partly #1864

## Summary of Changes
### Track Specific
- [X] FCU selected track mode.
- [X] FCU managed track mode.
- [X] FCU managed track mode with show heading.
- [X] FCU switch between heading to track and vice-versa.
- [X] PFD FMA track
- [x] PFD track heading tape
- [x] ND track mode

### FPA specific
- [x] FCU selected FPA mode.
- [x] FCU managed FPA mode.
- [x] FCU managed FPA mode with show datum
- [x] FCU switch between V/S to FPA and vice-versa.
- [x] PFD FMA FPA

### Generic
- [x] PFD FPV
- [x] PFD FPV Flight Director

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
Check the comments.

## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->
<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->
https://web.archive.org/web/20160302090326/http://williams.best.vwh.net/avform.htm#Wind
http://www.a320dp.com/A320_DP/nav-autoflight/sys-13.2.10.html
http://www.a320dp.com/A320_DP/nav-autoflight/sys-13.2.7.html
http://www.a320dp.com/A320_DP/nav-autoflight/sys-13.2.1.html
http://www.a320dp.com/A320_DP/nav-autoflight/sys-13.2.8.html
http://www.a320dp.com/A320_DP/nav-autoflight/sys-13.2.4.html

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): Theodore Messinezis

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

theomessin commented:
## Track Specific
- [X] FCU selected track mode.
- [X] FCU managed track mode.
- [X] FCU managed track mode with show heading.
- [X] FCU switch between heading to track and vice-versa.
- [X] PFD FMA track
- [ ] PFD track heading tape
- [ ] ND track mode

## FPA specific
- [ ] FCU selected FPA mode.
- [ ] FCU managed FPA mode.
- [ ] FCU managed FPA mode with show datum
- [ ] FCU switch between V/S to FPA and vice-versa.
- [ ] PFD FMA FPA

## Generic
- [ ] PFD FPV
- [ ] PFD FPV Flight Director

theomessin commented:
https://streamable.com/nodrn6

theomessin commented on A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_FCU.xml:
```diff
@@ -150,4 +150,98 @@
+
+		<Component ID="#NODE_ID#" Node="#NODE_ID#">
+			<UseTemplate Name="ASOBO_GT_Push_Button">
+				<LEFT_SINGLE_CODE>(L:A32NX_TRK_FPA_MODE_ACTIVE) ! (&gt;L:A32NX_TRK_FPA_MODE_ACTIVE)</LEFT_SINGLE_CODE>
```
Add to https://github.com/flybywiresim/a32nx/blob/master/docs/a320-simvars.md

theomessin commented on A32NX/en-US.locPak:
```diff
@@ -8,6 +8,8 @@
       "AIRCRAFT.UI_VARIATION": "Airbus",
       "COCKPIT.TOOLTIPS.AUTOPILOT_PANEL_HDG_TUNING_LEFT": "Adjust selected heading to the left (%((A:AUTOPILOT HEADING LOCK DIR:1,degrees))%!d!°)",
       "COCKPIT.TOOLTIPS.AUTOPILOT_PANEL_HDG_TUNING_RIGHT": "Adjust selected heading to the right  (%((A:AUTOPILOT HEADING LOCK DIR:1,degrees))%!d!°)",
+      "COCKPIT.TOOLTIPS.AUTOPILOT_PANEL_TRK_TUNING_LEFT": "Adjust selected track to the left (%((L:A32NX_AUTOPILOT_TRACK_SELECTED:1,degrees))%!d!°)",
```
Add to https://github.com/flybywiresim/a32nx/blob/master/docs/a320-simvars.md

theomessin commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FCU/A320_Neo_FCU.js:
```diff
@@ -336,7 +333,7 @@ class A320_Neo_FCU_Heading extends A320_Neo_FCU_Component {
             const simHeading = SimVar.GetSimVarValue("PLANE HEADING DEGREES MAGNETIC", "degree");
             Coherent.call("HEADING_BUG_SET", 1, simHeading);
-            const track = SimVar.GetSimVarValue("GPS GROUND MAGNETIC TRACK", "Radians") * 180 / Math.PI;;
+            const track = SimVar.GetSimVarValue("GPS GROUND MAGNETIC TRACK", "Radians") * 180 / Math.PI; ;
```
Remove extra `;`

theomessin commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FCU/A320_Neo_FCU.js:
```diff
@@ -296,34 +291,36 @@ class A320_Neo_FCU_Heading extends A320_Neo_FCU_Component {
-        const heading = _heading * Math.PI / 180;
         const trueAirspeed = SimVar.GetSimVarValue("AIRSPEED TRUE", "Knots");
-        if (trueAirspeed < 50) return _heading;
+        if (trueAirspeed < 50) {
```
Define this magic number in the class constructor.

theomessin commented:
Note to self: it's worth looking into fixing this as well (http://www.a320dp.com/A320_DP/nav-autoflight/sys-13.2.7.html)
```
If a heading is selected prior to pulling the selector, the airplane turns in the shortest direction to the selected heading. If a heading is selected after pulling the selector, the airplane turns to the new heading in the direction the selector is turned. If the selector is pulled during a turn, the airplane rolls out on the existing heading.
```

theomessin commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FCU/A320_Neo_FCU.js:
```diff
@@ -304,7 +349,25 @@ class A320_Neo_FCU_Heading extends A320_Neo_FCU_Component {
             this.backToIdleTimeout = 5;
```
Magic number, move to constructor. Also, not sure if this should be 45, as per http://www.a320dp.com/A320_DP/nav-autoflight/sys-13.2.8.html

theomessin commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/Airbus_FMA.js:
```diff
@@ -137,7 +137,7 @@ var Airbus_FMA;
             this.autoPilotVerticalSpeedHoldActive = Simplane.getAutoPilotVerticalSpeedHoldActive();
             this.autoPilotVerticalSpeedHoldValue = Simplane.getAutoPilotVerticalSpeedHoldValue();
-            this.autoPilotFlightPathAngle = Simplane.getAutoPilotFlightPathAngle();
+            this.autoPilotFlightPathAngle = SimVar.GetSimVarValue("L:A32NX_AUTOPILOT_FPA_SELECTED", "Degree");
```
Add to https://github.com/flybywiresim/a32nx/blob/master/docs/a320-simvars.md

theomessin commented:
FPA now works (kinda): https://streamable.com/ghk8cc

theomessin commented:
Note to self: use `Utils.Clamp` instead of checking values manually.

Benjozork commented on A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_FCU.xml:
```diff
@@ -150,4 +150,165 @@
+            <ANIMTIP_1_ON_CURSOR>TurnRight</ANIMTIP_1_ON_CURSOR>
+			<ANIMTIP_3>%((L:A32NX_TRK_FPA_MODE_ACTIVE, bool))%{if}Engage flight path angle%{else}Engage vertical speed%{end}</ANIMTIP_3>
+            <ANIMTIP_3_ON_CURSOR>DownArrow</ANIMTIP_3_ON_CURSOR>
+		</DefaultTemplateParameters>
+		<EditableTemplateParameters>
+			<AddParams>
+				<ANIM_NAME></ANIM_NAME>
+				<ANIM_NAME_PUSHPULL></ANIM_NAME_PUSHPULL>
+			</AddParams>
+		</EditableTemplateParameters>
+
+        <Component ID="#NODE_ID#" Node="#NODE_ID#">
+            <UseTemplate Name="ASOBO_GT_Knob_Infinite_PushPull">
+                <CLOCKWISE_CODE>
+                    (L:A32NX_TRK_FPA_MODE_ACTIVE, bool) 1 == if{
+                        (&gt;H:A320_Neo_FCU_AP_INC_FPA)
+                    } els{
+                        2 (&gt;K:AP_VS_VAR_INC)
+                        (&gt;H:A320_Neo_FCU_VS_INC)
+                    }
+                </CLOCKWISE_CODE>
+                <ANTICLOCKWISE_CODE>
+                    (L:A32NX_TRK_FPA_MODE_ACTIVE, bool) 1 == if{
+                        (&gt;H:A320_Neo_FCU_AP_DEC_FPA)
+                    } els{
+                        2 (&gt;K:AP_VS_VAR_DEC)
+                        (&gt;H:A320_Neo_FCU_VS_DEC)
+                    }
+                </ANTICLOCKWISE_CODE>
+                <PULL_CODE>
+                    (&gt;H:A320_Neo_FCU_VS_HOLD)
+                </PULL_CODE>
+                <PUSH_CODE>
+                    (&gt;H:A320_Neo_FCU_VS_ZERO)
+                </PUSH_CODE>
+                <CENTER_RADIUS>0</CENTER_RADIUS>
+            </UseTemplate>
+        </Component>
+	</Template>
 </ModelBehaviors>
```
Prefix template names with `FBW`, not `A32NX`.

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FCU/A320_Neo_FCU.js:
```diff
@@ -588,15 +664,55 @@ class A320_Neo_FCU_VerticalSpeed extends A320_Neo_FCU_Component {
+            SimVar.SetSimVarValue("L:A32NX_AUTOPILOT_FPA_SELECTED", "Degree", newFpa);
+        }
+    }
+    onFPAModeChanged(_newValue) {
+        if (_newValue) {
+            const verticalSpeed = Simplane.getAutoPilotVerticalSpeedHoldValue();
+            const angle = this.calculateAngleForVerticalSpeed(verticalSpeed);
+            SimVar.SetSimVarValue("L:A32NX_AUTOPILOT_FPA_SELECTED", "Degree", angle);
+        }
+    }
+    calculateVerticalSpeedForAngle(_angle) {
+        if (_angle == 0) {
+            return 0;
+        }
+        const _groundSpeed = SimVar.GetSimVarValue("GPS GROUND SPEED", "Meters per second");
+        const groundSpeed = _groundSpeed * 3.28084 * 60; // Now in feet per minute.
+        const angle = _angle * Math.PI / 180; // Now in radians.
+        const verticalSpeed = Math.tan(angle) * groundSpeed;
+        console.log("V/S for set FPA: " + verticalSpeed);
+        return (verticalSpeed < -6000) ? -6000 : (verticalSpeed > 6000) ? 6000 : verticalSpeed;
+    }
+    calculateAngleForVerticalSpeed(verticalSpeed) {
```
Can we add JSDoc comments for those functions ?

Benjozork commented on docs/a320-simvars.md:
```diff
@@ -344,3 +344,15 @@
+    - Degrees
+    - The selected track in the FCU
+
+- L:A32NX_AUTOPILOT_FPA_SELECTED
```
```suggestion
- A32NX_AUTOPILOT_FPA_SELECTED
```

theomessin commented:
Note to self: `The fixed Airplane Symbol is dimmed when TRK-FPA selected.` from http://www.a320dp.com/A320_DP/flight-instruments/sys-8.3.15.html. Can't find any images showing this...

Benjozork commented:
Please keep in mind that A320DP depicts EIS1 displays while NEOs use EIS2 displays. You'll have to find a reference for those.

theomessin commented:
Known bug: selected FPA will not be flown by A/P if autothrust is set and the aircraft is speeding above the selected speed. The same is true for V/S and this PR just uses the default V/S functionality under the hood. Fixing the underlying V/S will solve this.

theomessin commented:
FPV/FPD added (also kinda): https://streamable.com/8x61tk

Benjozork commented on A32NX/ModelBehaviorDefs/A32NX/Interior/A32NX_Interior_FCU.xml:
```diff
@@ -150,4 +150,160 @@
+			</AddParams>
+		</EditableTemplateParameters>
+
+        <Component ID="#NODE_ID#" Node="#NODE_ID#">
+            <UseTemplate Name="ASOBO_GT_Knob_Infinite_PushPull">
+                <CLOCKWISE_CODE>
+                    (L:A32NX_TRK_FPA_MODE_ACTIVE, bool) 1 == if{
+                        (&gt;H:A320_Neo_FCU_AP_INC_FPA)
+                    } els{
+                        2 (&gt;K:AP_VS_VAR_INC)
+                        (&gt;H:A320_Neo_FCU_VS_INC)
+                    }
+                </CLOCKWISE_CODE>
+                <ANTICLOCKWISE_CODE>
+                    (L:A32NX_TRK_FPA_MODE_ACTIVE, bool) 1 == if{
+                        (&gt;H:A320_Neo_FCU_AP_DEC_FPA)
+                    } els{
+                        2 (&gt;K:AP_VS_VAR_DEC)
+                        (&gt;H:A320_Neo_FCU_VS_DEC)
+                    }
+                </ANTICLOCKWISE_CODE>
+                <PULL_CODE>
+                    (&gt;H:A320_Neo_FCU_VS_HOLD)
+                </PULL_CODE>
+                <PUSH_CODE>
+                    (&gt;H:A320_Neo_FCU_VS_ZERO)
+                </PUSH_CODE>
+                <CENTER_RADIUS>0</CENTER_RADIUS>
+            </UseTemplate>
+        </Component>
+	</Template>
```
Indentationa

Benjozork commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FCU/A320_Neo_FCU.js:
```diff
@@ -508,23 +597,27 @@ class A320_Neo_FCU_VerticalSpeed extends A320_Neo_FCU_Component {
+                Coherent.call("AP_VS_VAR_SET_ENGLISH", 1, verticalSpeed);
+                Coherent.call("AP_VS_VAR_SET_ENGLISH", 2, verticalSpeed);
+            }
+            // const targetVerticalSpeed = Simplane.getAutoPilotVerticalSpeedHoldValue();
+            // if (deltaAltitude * targetVerticalSpeed < 1) {
+            //     this.currentState = A320_Neo_FCU_VSpeed_State.Idle;
+            //     this._enterIdleState();
+            //     this.forceUpdate = true;
+            // }
```
Remove commented code

Benjozork reviewed (approved):
LGTM

danbw9 commented:
Great idea with this feature. Though before I could properly test it out, I found a major issue. Whenever I am in TRK/FPA mode, and try to change the FPA UP and DOWN, it changes the TRK value on FCU and not the FPA value. So the FCU knob is linked to the wrong window. Could you check what this is? And also rebase on the latest master. Thanks!

starvingjavi commented:
Category: QA Tester
Name: Javi#2557
Date of testing: 21/12/2020
Version of the sim: 1.11.7
PR Tested: #2389 
Object of test: FCU TRK functionality

Testing technique: Full flight from SACO to SAME while testing at different points in flight the "bird", specially to follow a 3° descent for landing
Steps you've made to test the Pull Request:
1. Start flight as normal
2. Test between normal FD and TRK FPA
3. Continue flight
4. Perform manual flight with TRK FPA


Flight notes (If required): Performance has decreased significantly since 0.5.0. I used to get ~30 FPS and now I get 15FPS.

Normal flight otherwise. Route was IRAVO1 ALDEX UA307 SALBO SALB1Q

Issues (If Any): No
Overall Rating: Good
Conclusions of the testing: The FPA/TRK mode works as expected and according to the documentation provided. Was able to follow a 3° descent with it for landing. PR works as intended. Performance of the simulator has decreased considerably

theomessin commented:
@danbw9 The V/S knob has been corrected. Apologies, I messed it up when I was renaming the template names.

Regarding performance, @starvingjavi can you confirm whether you see the FPS drop when you're in TRK/FPA mode, or generally. My code only runs if in TRK/FPA mode is on so that's where I'm expecting an FPS drop. If you're seeing an overall drop, that's different and will have to be investigated. 

I added some checks to recalculate heading for track and vertical speed for angle less often, let me know if this helps.

starvingjavi commented:
No performance loss when going to TRK/FPA mode. It must be something else then.

wpine215 commented:
> 
> 
> FPV/FPD added (also kinda): https://streamable.com/8x61tk

Is it possible to animate the FPV/FPD to slowly update when the TRK/FPA is changed on the FCU, instead of jump in steps?

theomessin commented:
@wpine215, not sure. It's just the FPD `y` value that jumps in steps when the FPA changes. Everything else is smooth. I think it's okay until we get our own AP. AFAIK people don't really use the FPD, just the FPV for non-precision approaches. 

I tried using CSS transitions but that didn't work. Only other alternative I can think of is to "animate" the `y` value in FPD's refresh method using code (personally I don't think it's worth it).  Maybe someone else can pick this bit up?

MisterChocker commented:
We currently use some smoothing on the pfd to make things have a nice transition. I can show you examples in a couple of hours.

danbw9 commented:
Category: QA Test
Name: danbw9#7028
Date of testing: 23/12/2020
Version of the sim: 1.12.13.0
PR Tested: #2389
Object of test: Testing the FCU TRK/FPA Mode

Testing technique: Raw data flying, multiple approaches (VOR/NDB/LOC) and full flight competed from ESMS to ESGG + observe the effect of wind and environmental conditions on the functionality

Steps you've made to test the Pull Request:

Check that the included items listed above are actually in the PR. All correct. 

Initial check of the PR changes whilst spawning on RWY. Confirm the functionality is correct on FCU and on FMA for that phase of flight - when spawned in air the indications are fine. On the ground and after a landing with TRK FPA mode, the mode does not display anymore on the PFD. 

Confirmed that the ND Track mode is not implemented yet - as is unchecked above. 

I did some flying around vectoring myself with AP on and I noticed a misalignment between the commanded TRK and the FPV. It appears to have 0 Angle of bank and to have settled but still 1-2 degrees off. Over time it seems to get better. This could be to do with the 1 degree off for the Cyan bug (see second screenshot below) - It is on ND though which is not correct yet, but I noticed this is also an issue before the PR with just HDG. 

Throughout the testing I noticed a crosswind throws off the indications - shown in raw data FDs off. The PFD TRK tape (once on the desired FCU setting), is always centered in the PFD when it should be aligned with the FPV when on the TRK. See SS1 directly below.

SS1
![Screenshot (1057)](https://user-images.githubusercontent.com/1753445/103059945-dc6dc080-45a6-11eb-800e-4b2826e2fcc7.png)



The following approaches were attempted:
- LEBL VOR RWY02
- ENVA LOC RWY09
- UAAA VOR DME RWY05L
- EGHH NDB DME RWY26

Tryout NPA - LEBL VOR RWY02. Platform ALT 1500 ft. Descent point 5 DME PRA VOR. Initiate FPA -3.2° at 5.3 DME - awkwardness of the FCU/AP interaction means have to fiddle a lot beforehand. Did check altitude/distance checks at 4, 3 & 2 DME and the FPA worked all fine. TRK line set to inbound course and using ROSE VOR mode there were no issues. I repeated the same procedure for a stabilized non precision approach for the other approaches and largely it's very just the ND TRK is not reliable and can make it appear as if it's not established on the course. 

Also flew a raw data circuit at EGNX. No issues. Though as mentioned above, the crosswind created some confusion over the indications for TRK FPA on the PFD versus the ND. 

Issues (If Any): A few. TRK FPA functions disappear when on ground. FPV could be more bold in font as it is currently not so easy to see amongst the other lines/colours.  TRK tape on PFD is a little off as described above.
Overall Rating: Needs improvement
Conclusions of the testing: Overall great. Mostly LGTM. Confirmed that the two issues that were reported previously like the performance decrease as well as V/S knob controlling the HDG on FCU - do not occur anymore. TRK PA visuals on PFD should be there on the ground and not only in the air. The visuals also could be more clear and the TRK tape should be fixed. 

This is quite an essential feature to have with the mod with Non precision approaches, so thanks for your work on this

SS2
![Screenshot (1052)](https://user-images.githubusercontent.com/1753445/103056660-ce677200-459d-11eb-90b0-bde560fe2a4f.png)
SS3
![Screenshot (1051)](https://user-images.githubusercontent.com/1753445/103056794-361dbd00-459e-11eb-8ca7-940780483a21.png)

wpine215 commented:
> 
> 
> We currently use some smoothing on the pfd to make things have a nice transition. I can show you examples in a couple of hours.

Any update on this?

theomessin commented:
@danbw9 I haven't touched the FMA so any bug there should be present in HDG mode as well. None of your SS have FD on, so make sure that's on (otherwise you won't get see anything on the FMA). 

> Throughout the testing I noticed a crosswind throws off the indications - shown in raw data FDs off. The PFD TRK tape (once on the desired FCU setting), is always centered in the PFD when it should be aligned with the FPV when on the TRK. See SS1 directly below.

Not too sure what you mean by the PFD tape issues. When in TRK/FPA mode, the tape is as follows:
![Untitled](https://user-images.githubusercontent.com/7229472/103137866-dc63e280-46c4-11eb-9221-a723bb6d6302.png)

I agree that the FPV is on the thin side, I'll look into making it bold. I haven't coded the ND TRK mode yet - I'd like some references as to how this looks in the real aircraft. The same applies with the PFD, I'm still not sure how the real aircraft TRK/FPA mode affects the PFD so I had to go with what I could find.

> I did some flying around vectoring myself with AP on and I noticed a misalignment between the commanded TRK and the FPV. It appears to have 0 Angle of bank and to have settled but still 1-2 degrees off. Over time it seems to get better. This could be to do with the 1 degree off for the Cyan bug (see second screenshot below) - It is on ND though which is not correct yet, but I noticed this is also an issue before the PR with just HDG.

Yeah, this is a known issue that has to do with the default AP in heading mode. TRK just sets an underlying heading and the AP just can't fly that heading accurately. It's not worth fixing this as we're getting our own custom AP (which will replace this PR) at some point.

danbw9 commented:
> haven't touched the FMA so any bug there should be present in HDG mode as well. None of your SS have FD on, so make sure that's on (otherwise you won't get see anything on the FMA).

Thanks for the response. And yeah I was testing with both FD off and on. Raw data flying with FD off is where I noticed the bug with the wind and I was testing that out. 

> Not too sure what you mean by the PFD tape issues. When in TRK/FPA mode, the tape is as follows:

When on the commanded track but with a crosswind (in case of SS1 wind from the left), the blue vertical line for TRK indicator should be aligned with the green FPV symbol and not centered in the PFD display - unless 0 X wind component. See SS1 for a little reference.

> I agree that the FPV is on the thin side, I'll look into making it bold. I haven't coded the ND TRK mode yet - I'd like some references as to how this looks in the real aircraft. The same applies with the PFD, I'm still not sure how the real aircraft TRK/FPA mode affects the PFD so I had to go with what I could find.

With regards to this, I can do my best to find references for you although at first glance they are not easy in the manuals or via YouTube.

tareksabet commented:
Name: tarek_sabet#7946
Date of testing: Current date 03/01/2021
Version of the sim: 1.12.13.0
PR Tested: #2389 
Object of test: Testing TRK/FPA Mode 
Testing technique: Try-out (multiple tests)
 
## Steps you've made to test the Pull Request:
* Did multiple Landings on non precision approaches 
*Did 1 of them a VOR landing at HECA 05C Managed Selected ( selected heading, Managed Altitude )
*Did 2 VOR landings at HESH 04R Managed Selected ( selected heading, Managed Altitude ) 2nd landing with heavy crosswind 
*Did 1 VOR landing at HESG 33 Managed Managed  ( Managed heading, Managed Altitude )

## Issues (If required):
*I have the same issue as in Dan's report about the TRK on ND but you didn't work on it yet also the FPA and FPV should be thicker as in the video I mentioned 

## Overall Rating: To be Improved  
### Conclusions of the testing:
* with your PR I can finally do non precisions approaches everything is good despite the 2 issues I mentioned but other than that, I love it. 
Thanks for the great job you did.

### Video:
* The video below is a video that I recorded in an A320 sim which I performed an approach HECA VOR 05C.


https://youtu.be/kCx4WuJ2UDk

tareksabet commented:
Name: tarek_sabet#7946
Date of testing: Current date 03/01/2021
Version of the sim: 1.12.13.0
PR Tested: #2389 
Object of test: Testing TRK/FPA Mode 
Testing technique: Try-out (multiple tests)
 
## Steps you've made to test the Pull Request:
* Spawned at different Airports:
1- HECA
2- HESH 
3- HESG 
(All airports are in Egypt) 
* Performed multiple VOR landings at these airports. 

## Flight report :
Flight flight was at HECA, I did 1 VOR landing on runway 05C with Managed Selected modes ( selected heading, Managed Altitude ).
Then second flight, I did 2 VOR landings on runway 04R Managed Selected ( selected heading, Managed Altitude ) 1st one was clear weather and the 2nd one was with some turbulence and a 20Kt crosswind.
The third flight, I did 1 VOR landing on runway 33 Managed Managed  ( Managed heading, Managed Altitude ).

All these landings were respecting the charts and thy are listed below.

## Issues (If required):
* I had the same issues as in Dan's report about the TRK on ND but as mention you didn't work on it yet also, the FPA and FPV should be thicker as in the video I uploaded below.

## Overall Rating: To be Improved  
### Conclusions of the testing:
* with your PR I can finally do non precisions approaches everything is good despite the 2 issues I mentioned but other than that, I love it. 
Thanks for the great job you did.

### Video:
* The video below is a video that I recorded in an A320 sim which I performed an approach at HECA landing on rwy 05C using VOR.


https://youtu.be/kCx4WuJ2UDk

###Approache Charts used :
* HECA VOR landing rwy 05C

![image](https://user-images.githubusercontent.com/70240330/103471861-60396100-4d8e-11eb-9cf4-b1a24be6397f.jpeg)

* HESH VOR landing rwy 04R 

![image](https://user-images.githubusercontent.com/70240330/103471865-692a3280-4d8e-11eb-9813-ad21a2405ae4.jpeg)

* HESG VOR landing rwy 33

![image](https://user-images.githubusercontent.com/70240330/103471888-bad2bd00-4d8e-11eb-9212-eea06582369a.jpeg)

TwitchyShell commented:
**Category:** QA Tester
**Name:** Тайга Мөңгүн#2000
**Date of testing:** 2021-01-03
**Version of the sim:** 1.12.13.0
**PR Tested:** #2389 
**Object of test:** Testing TRK/FPA Mode

# **Testing technique:**
Full line flight, 2 sectors, UHSS UHSH, UHSM UHSS involving 2LOC approaches. Meteoblue live weather used.

Secondary airwork flight conducted in turbulence and crosswind at UHSM, Meteoblue live weather used as well.

## **Steps you've made to test the Pull Request:**
### Line flights:
1. Preflight as normal
2. TO and departure in TRK FPA but managed
3. Climb in managed nav and selected track
4. Normal cruise
5. Descent as normal
6. Approach in TRK and FPA
7. FD off when visual, manual flight to landing with reference to FPV

### **Airwork**
1. Spawn airbourne
2. Test behaviour of FPD in turbulence in selected TRK and FPV and record video samples

# **Flight notes:**
### **Line flights:**
Powered up, selected TRK FPA, proper behaviour on the ground. Completed on ground actions at UHSS. Take off, FPV and FPD appear and behave as expected. Initial climb behave as expected.

Further climb segment, selected manual TRK, **observations regarding ND TRK bug by danbw9 and tareksabet also seen** on these flights and **acknowledged as not having been implemented yet**. Under manual TRK, PFD FMA changes from "THR CLB, CLB, NAV" to "SPEED, blank, TRK", which is expected behaviour. Quick airwork exercises under 20 knot crosswind have manual TRK behaving satisfactory and as expected. When the HDG V/S - TRK FPA pushbutton is selected, FCU HDG/TRK window swaps, and the window displays the digit readout as expected; a different value for HDG and TRK.

Cruise and descent nil comment.

First approach was UHSH RWY 31 2LOC approach. Turbulence encountered. **FPD appeared very sensitive and jittery**, it attempted to correct immediately to any turbulence bumps. For FPD pitch (FPA), please refer to below section in airwork. Approach conducted until mins, then FPD removed. **FPV behaves very well, accurate and is steady**, does not exhibit jitter issues like FPD. 

Upon touchdown, FPV disappear with as expected.

Second line flight conducted with similar results. No turbulence encountered on this flight. FPD did not exhibit jitter.

### **Airwork:**
Spawned at UHSM and conducted quick airwork in manual flight to verify certain FPD behaviours in known turbulence and crosswinds. Jitters observed again on FPD, as demonstrated in first two excerpts in video sample.

Conducted FPA descents and climbs, -3.0 deg, -2.6 deg and +7.0 deg selected FPA modes used. **Observed jitter in the pitch axis of FPD**, likely FPD attempting to compensate for turbulence which is incorrect behaviour. **Attempted FPA +7.0 deg from off pitch position (FPV ~-2 deg), observed FPD commanding +11.0 deg, this is incorrect behaviour.** FPD should always remain at value of selected FPA, regardless of conditions, (i.e. FPA +7.0 deg, FPD will always stick at +7.0 deg position, aircraft will then attempt to achieve FPD command.)
Reference: Anecdotal observations from tareksabat and his included YouTube video, towards the end of his video, flight in turbulence observed.

I do acknowledge this is an associated issue with the underlying FD implementation.

AWK video sample: https://youtu.be/jATtvraLftI

# **Issues (If Any):** 
* FPD jitter under turbulence
* ND TRK bug feature incomplete
* Symbology thickness/appearance

### **Overall Rating:** Good

# **Conclusions of the testing:** A very pleasant and functional implementation. Good behaviour on PFD and FCU, just ND TRK bug implementation required. Reason for this is because some non precision approaches (NDB, 2LOC) do not have CRS that you can assign (unlike VOR or LOC), which makes ND TRK indications are important for these operations.

The FPD jitter issues and behaviour are relatively minor issues, they still generally perform their function as needed, and I acknowledge some issues might be related to the split axis FD implementation.

Thank you for implementing this feature as it provides me personally a lot of enjoyment and ease in conducting 2LOC operations.

theomessin commented:
Hi all,

I've added track mode for the ND (using the video @tareksabet gave us). I've also made the FPV and FPD thicker. @TwitchyShell I've fixed the jitter for selected mode only (if you've set a specific FPA). For managed mode we have to use the default FD and there's no way to fix this without a performance hit (we could try to smooth the data but it seems a bit excessive for this stage).

![Capture](https://user-images.githubusercontent.com/7229472/103625698-3cfbe680-4f33-11eb-9210-14e747465a41.PNG)

theomessin commented:
> When on the commanded track but with a crosswind (in case of SS1 wind from the left), the blue vertical line for TRK indicator should be aligned with the green FPV symbol and not centered in the PFD display - unless 0 X wind component. See SS1 for a little reference.

Fixed!

danbw9 commented:
Thanks a lot and great work @theomessin. 

I just did another test flying an approach and it works flawlessly as before. Able to fly a NPA procedure correctly. 

Also did some airwork with a 20 knot crosswind component - to test that the TRK indication in the centre of the PFD has been corrected. Coordinated this in discord VC with Theodore for it to be fixed. 

Gone over the above issues described in the other tests. Symbology of TRK FPA mainly the FPV has been improved a lot and is more accurate now. No Jitters experienced. 

LGTM and green to merge into master. Very cool feature

MisterChocker commented:
> Thanks a lot and great work @theomessin.
> 
> I just did another test flying an approach and it works flawlessly as before. Able to fly a NPA procedure correctly.
> 
> Also did some airwork with a 20 knot crosswind component - to test that the TRK indication in the centre of the PFD has been corrected. Coordinated this in discord VC with Theodore for it to be fixed.
> 
> Gone over the above issues described in the other tests. Symbology of TRK FPA mainly the FPV has been improved a lot and is more accurate now. No Jitters experienced.
> 
> LGTM and green to merge into master. Very cool feature.

Did you also ensured when on heading the heading indication is still correct and does not mirror the track behavior?

danbw9 commented:
> > Thanks a lot and great work @theomessin.
> > I just did another test flying an approach and it works flawlessly as before. Able to fly a NPA procedure correctly.
> > Also did some airwork with a 20 knot crosswind component - to test that the TRK indication in the centre of the PFD has been corrected. Coordinated this in discord VC with Theodore for it to be fixed.
> > Gone over the above issues described in the other tests. Symbology of TRK FPA mainly the FPV has been improved a lot and is more accurate now. No Jitters experienced.
> > LGTM and green to merge into master. Very cool feature.
> 
> Did you also ensured when on heading the heading indication is still correct and does not mirror the track behavior?
> But I assume by looking at the code it should work just fine.

Good thought. I did try that and it's all good

Benjozork commented:
Thank you to the author everyone who helped shape this PR! Merging!
